### ISSUE TITLE
Logout all sessions does not work well with new store/ infinispan external cache

### ISSUE THREAD

**Issue by hmlnarik** [keycloak/keycloak#15592]
### Discussed in https://github.com/keycloak/keycloak/discussions/15168

<div type='discussions-op-text'>

<sup>Originally posted by **kkcmadhu** October 26, 2022</sup>
Hi, i enabled new store and storing session details in external infinispan cache, using hotrod.

I was testing/perf testing few basic use cases, and i see that over the time, the time taken to get token increases  ( i am using basic auth), i also noticed that the number of session in keycloak kept increasing (which is expected), i wanted to try out logut All to see if the response time comes down, but unfortunately the "logout All sessions" is not having any effect at all.
When i click logout all, i see this error in my keycloak pod 

```
java.util.concurrent.CompletionException: org.infinispan.commons.CacheException: org.hibernate.search.util.common.SearchException: HSEARCH600067: Unable to delete all entries matching query '*:*': Lock held by this virtual machine: /opt/infinispan/server/data/kc.HotRodAttributeEntity/write.lock
Context: index 'kc.HotRodAttributeEntity'
org.infinispan.commons.CacheException: org.hibernate.search.util.common.SearchException: HSEARCH600067: Unable to delete all entries matching query '*:*': Lock held by this virtual machine: /opt/infinispan/server/data/kc.HotRodAttributeEntity/write.lock
Context: index 'kc.HotRodAttributeEntity'
org.hibernate.search.util.common.SearchException: HSEARCH600067: Unable to delete all entries matching query '*:*': Lock held by this virtual machine: /opt/infinispan/server/data/kc.HotRodAttributeEntity/write.lock
Context: index 'kc.HotRodAttributeEntity'
org.apache.lucene.store.LockObtainFailedException: Lock held by this virtual machine: /opt/infinispan/server/data/kc.HotRodAttributeEntity/write.lock
	at org.infinispan.client.hotrod.impl.protocol.Codec20.checkForErrorsInResponseStatus(Codec20.java:323)
	at org.infinispan.client.hotrod.impl.protocol.Codec20.readHeader(Codec20.java:168)
	at org.infinispan.client.hotrod.impl.transport.netty.HeaderDecoder.decode(HeaderDecoder.java:139)
	at org.infinispan.client.hotrod.impl.transport.netty.HintedReplayingDecoder.callDecode(HintedReplayingDecoder.java:94)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:279)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.handler.timeout.IdleStateHandler.channelRead(IdleStateHandler.java:286)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
	at io.netty.channel.epoll.AbstractEpollStreamChannel$EpollStreamUnsafe.epollInReady(AbstractEpollStreamChannel.java:795)
	at io.netty.channel.epoll.EpollEventLoop.processReady(EpollEventLoop.java:480)
	at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:378)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:986)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
2022-10-26T11:47:18.470674849Z
```
</div>

**Comment by mhajas** [comment]
I played with this a little bit and I am still not able to reproduce the slowness when getting tokens. 

However, I was able to reproduce the slowness with logout all sessions. I created 10000 sessions with KC 19 and then tried to press logout all and it has basically no effect (the request takes toooo long, I stopped it after some time but it will probably finish eventually),

I am not able to reproduce the exception though, @kkcmadhu is it possible you pressed log out all more than once? I would say it would explain why the exception is visible.

The issue with logging out of all sessions is that we are removing also all client sessions for each user session. If we have 10000 user sessions with 10000 client sessions (1:1) for one user, then when removing all of them, we need to load all user session ids and then remove all client sessions where user-session-id is in that list. This may even hit some limits for Ickle query length.

See the code:
https://github.com/keycloak/keycloak/blob/b8e01596f65b13f0ec5607e42393f573059ad26d/model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/userSession/HotRodUserSessionTransaction.java#L159-L169


I tested with version 19 and also the current main. The logout all issue is present in both, I tested with 1000 user sessions and logout all takes 13s for 19 and 10s for main.

I see two ways how to fix that:
1. simple one - don't remove the client session and let Infinispan lifespan evict them (I believe this should be just about removing one line from the code snippet above)
2. make some optimization for removal, e.g. storing user-id in the client session and then do some criteria magic to get user id out of the criteria within the method above and remove sessions using query (WHERE user-id = "") which should be much faster.
